feat!: Introduce OCI multi-image index manifests

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

> **Warning**
>  This PR contains breaking changes to the OCI package manager. Upon release, all packages built on the OCI package manager will need to be re-packaged!

This large refactor reworks the OCI package manager to assume the use of the OCI image specification's index such that we can represent multiple manifests within a single canonical referencable name.  This means that we can build multiple unikernels with the same name.  This makes sense for multi-target projects which build the same application but target different hardware/platform vendors.

The structure of the index can be found [here](https://github.com/opencontainers/image-spec/blob/main/image-index.md) and in this implementation we itemized each manifest in the index in the same way. When we use the `Catalog` method now, we are able to query a remote registry based on the selected OS (platform) and architecture as well as any embedded KConfig options (if they are embedded).

When packaging multiple targets of a single application, the resulting package will be listed for all unique targets:

```
TYPE  NAME        VERSION  FORMAT  DIGEST   PLAT
app   helloworld  latest   oci     e8be0a9  qemu/x86_64
app   helloworld  latest   oci     a6802f4  qemu/arm64
```

Use cases include selecting appropriately based on the host's architecture as well as feature selection via embedded KConfig options.

Notable additional improvements which are incorporated in this pull request include:
- Better offline mode in each package manager: all internal package managers were optimized to reduce instantiation time which makes a noticeable impact when using `kraft` in offline mode;
- Better outputs in `kraft pkg ls` which now has dynamic column data information;
- Better prompts in `kraft pkg` for selecting and packaging individual targets from an application project;
- `kraft pkg prune` has been renamed to `kraft pkg rm` and now works for all underlying implementations (per-manifest type for libraries and components, and directory and containerd for OCI packages);
- Better handling of insecure authenticated OCI registries; and,
- Natively selecting the host platform and architecture in `kraft pkg pull` when no additional flags are provided.

GitHub-Closes: #704
GitHub-Supersedes: #719
GitHub-Depends-On: #874
GitHub-Depends-On: #878

craciunoiuc committed:
feat(packmanager): Add options to specify package characteristics

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung committed:
refactor!: Remove unused target attribute

This commit removes the field method "format" from the target
specification.  The usage was intended to be part of the Kraftfile
specification, wherein a user would specify the format of the target
via:

```yaml
targets:
- name: example
  architecture: x86_64
  platform: qemu
  format: oci
```

However, this was never realised and introduces some complications for
now that are not resolved, specifically to do with the name of the
target which conflicts with internal usage (by both the `unikraft/`
package and the relevant package manager).  In this commit, refactor
the `kraft pkg` subcommand to ignore this attribute

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Converet `cmdfactory.NewEnumFlag` to generic

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix: Use implicit knowledge about the default manifest to speed up init

In most cases, the user will have the default manifest location.  It is
not necessary to check if this is compatible since it is inherently
compatible.  Instead of performing this check, which ultimately makes a
remote connection call, add the manifest and do not run `IsCompatible`.
To make this possible, export the constant representing default
location.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Invert package manager query attribute `NoCache` to `Update`

In contrast to "no cache", the term "update" is used to declare that the
package manager should perform a remote check when making the query
instead of using local cache.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Remove `pack.WithPullArchitecture` & `pack.WithPullPlatform`

These two method options are redundant since they are already part of
the package and are determined pre-emptively during the querying of the
catalog.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Remove autosave

This commit removes the "auto-save" ability from the OCI image
abstraction.  This is done for two reasons: a). it is not used anywhere
in the KraftKit codebase and b). it has an affect on containerd's
garbage collector.  Image components must be saved in a particular order
in order to ensure they are not removed by the garbage collector.  To
prevent this, removing autosave reduces complexity and ensures that that
during the image's `Save` method, items are performed in the correct
order.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Remove unused working directory

This variable (and by extension option to set the workdir) was never
used.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Rename `Image` to `Manifest`

This structure better describes the OCI Image Specification of a
"Manifest" than an "Image".  The image, in the context of the OCI spec,
only holds metadata associated with the runtime (.e.g entrypoint, env)
whereas the Manifest encompasses both this AND the rootfs (i.e. layers).

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Rename `manifestDesc` to `desc`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Use KraftKit's native auth list internally

This commit refactors the OCI package such that it uses KraftKit's
native list of auths instead of type-converting to the equivilant type
which overall simplifies the code.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Set and use the embedded reference to the manifest

The manifest is set at the end, remove this unnecessary assignment and
begin by instantiating the internal structure.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Use pointers on internal OCI spec structures

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(handler): Remove unused attribute and rename structure

The OCI Image Manifest is the proper name for this structure.  Rename
appropriately.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Correct method comment regarding inheritance

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce dynamic column information for packages

This commit reworks the idea of "package metadata" by a). returning
relevant information in the OCI package for the `Metadata` attribute and
b). introduces a `Columns` attribute which is subsequently used by
`kraft pkg ls`.  This results in a richer output.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci): Implement `ResolveManifest` and `DeleteManifest` handlers

This commit reworks the OCI package's `Delete` method.  In its previous
form, it assumed the directory implementation's handler which is
incorrect since the OCI package manager relies on an underlying
implementation for storing the artifacts of the OCI image spec.  To
accomplish this, containerd and the directory handler were updated with
relevant methods `ResolveManifest` and `DeleteManifest` interface
methods.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor!: Remove `pack.WithPullVersion`

This method options are redundant since they are already part of the
package and are determined pre-emptively during the querying of the
catalog.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(manifest): Introduce per-provider `DeleteManifest` method

Towards allowing each provider to determine how to delete the artifact
with which it represents, introduce the `DeleteManifest` provider
interface method.  Since each provider understands the implementation of
handling its artifact, use this newly introduced method `DeleteManifest`
in the package's `Delete` method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Request `WithUpdate` in the internal update method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix: Let each package manager handle "pruning" packages properly

This fix reworks the `kraft pkg prune` subcommand by letting the package
manager handle the deleting of a package by their respective underlying
implementations.  To this end, the following changes were done:

- `kraft pkg prune` was reworked to iterate through each package
  manager;
- The manifest package manager never actually deleted the package from
  the list of indexes, it only removed the manifest sources and was
  specific to the manifest provider and not the underlying provider;
- The manifest package manager should not fail-fast when "probing" each
  enrty based on the provider, instead the error is bubbled up as a
  debug message which allows the cursive method
  `findManifestsFromSource` to return the list of manifests,
  specifically NOT including newly deleted entries;
- The OCI package manager was specific to the directory provider and not
  the underlying handler (whether directory or containerd).

Additionally, the unused `version` argument on the `Package` interface
was removed, since this is embedded within the package itself.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(prune): Add copyright preamble

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor: Rename `kraft pkg prune` to `kraft pkg rm`

This is more inline with existing subcommands for `kraft`.  Also rename
the package manager's `Prune` method to `Delete` since "prune" implies
all and the method accepts query parameters to allow for selecting all
or a subset.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(packmanager): Simplify `Delete` method

Unnecessary errors are thrown resulting in early returns.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Only request manifest from URL on update

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Sort list of manifests in the index before saving

This makes diffs easier to understand when changes occur to said index.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(manifest): Improve initialization speed

This commit uses inherent knowledge about KraftKit to speed up the
initialization of the manifest package manager by preventing a remote
checks via an update.  If sources have been added to the manifest, they
are likely valid.  Instead, run the update only if a request to update
is made.  When no sources in the internal manifest cache, then we
iterate over each manifest in the config and save appropriately.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix: Improve performance & reliability of determining Git(Hub) provider

GitHub is natively a Git source, first check whether the remote source
is Git before checking it is from GitHub.  Additionally, use the host
name to quickly error out if the provided source does not originate from
`github.com`.  Note, this now no longer accounts for GitHub Enterprise,
however this use case has not yet presented itself.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(packmanager): Increase log verbosity

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(handler): Introduce `PullDigest` method

This method allows for pulling an object based on its digest.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(handler): Remove unused `DescriptorResolver` interface method

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Correctly handle saving the manifest

This commit fixes an issue where the composition of a manifest is not
saved correctly by containerd.  Whilst the change occurs at a level
higher than the containerd handler, the order here is important since
this is recognised by containerd and its garbage collector.  Ensure that
the manifest is first saved and then its entities such that the
previously unreferrenced entities are known before they are saved.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Introduce new constructor `NewManifestFromDigest`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Sort the KConfig "OS Features" before saving

This increases readability but also helps with creating repeatable
digests as maps are unsorted and would result in different checksums.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(handler): Introduce `SaveDescriptor` method

`SaveDescriptor` accepts an optional name reference which represents
descriptor (but this is not always necessary and can be left blank if
the descriptor is unnamed, e.g. an untagged config, a layer, etc) as
well as an io.Reader which is prepared to pass in the byte slice of
the descriptor.  An optional progress method callback can be provided
which is used to deliver the progress of writing the descriptor by
the implementing method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(handler)!: Output `ListManifests` as map with digest as key

Additionally introduce a generic helper method in containerd's
implementation which is able to perform the same operation based on
different inputs of mediatype and interface structures.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci)!: Deprecate `SaveDigest` in favour of `SaveDescriptor`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Introduce compatibility check helper methods

These methods determine if the provide source is compatible with
KraftKit, thus abstracting away the check in the attempt to reduce the
noise across the codebase for the same check.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(oci):! Implement and use OCI Image Indexes for mutli-images

This large refactor reworks the OCI package manager to assume the use of
the OCI image specification's index such that we can represent multiple
manifests within a single canonical referencable name.  This means that
we can build multiple unikernels with the same name.  This makes sense
for multi-target projects which build the same application but target
different hardware/platform vendors.

The structure of the index can be found at [0] and in this
implementation we itemized each manifest in the index in the same way.
When we use the `Catalog` method now, we are able to query a remote
registry based on the selected OS (platform) and architecture as well as
any embedded KConfig options (if they are embedded).

Use cases include selecting appropriately based on the host's
architecture as well as feature selection via embedded KConfig options.

Notable additional changes include:

- Removal of `FetchImage` which was replaced by `PullDigest`;
- Removal of `PushImage` which was replaced by `PushDescriptor`;
- Removal of `ResolveImage` which was replaced by `DigestExists`;
- Reworking the `Catalog` structure to parse indexes containing
  manifests; and,
- Introduction of `NewIndex`, `NewIndexFromSpec` and `NewIndexFromRef`
  which are used to populate an internal index representation from
  different sources.

[0]: https://github.com/opencontainers/image-spec/blob/main/image-index.md

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(internal): Import file locking packages verbatim

This commit copies in file locking packages from Go's internal
source[0] code verbatim.  The only adjustment made to the packages
are the removal of test cases which relied on additional imports
as well as changing the import path.

These packages are to be used when reading and writing files that
are performed asynchronously.

Finally, these files were passed through `gofumpt` to make their
syntax compliant with KraftKit's style guide.

[0]: https://github.com/golang/go/tree/go1.21.3/src/cmd/go/internal/lockedfile

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Correctly handle access to authenticated insecure registries

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(confirm): Introduce simple TUI query (yes/no) prompt helper

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(selection): Introduce CLI helper utility for selection prompts

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Introduce ability to overwrite or merge during packaging

This commit introduces a "merge strategy" system and allows the user to
decide during the invocation of `kraft pkg` either by CLI or via TUI
prompt whether to overwrite or merge into an existing package.  For now
the OCI package manager package implements this and enables users to
construct unikernels sequentially whilst still retaining the same image
name.  Further, the user can pull an existing image and add new
kernels.

To make this possible, the packmanager package was updated to introduce
a native option `PackMergeStrategy` along with 3 merge strategies:

- StrategyExit: Exit on existing/conflict;
- StrategyMerge: Merge new targets with an existing package;
- StrategyOverwrite: Totally remove the existing package and set the
  contents to the new targets.

Additionally, the OCI package was updated to accommodate for this
functionality and iterates across each existing manifest and checks the
the metadata (platform, architecture and set KConfig options) to
determine whether to overwrite.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Add more verbose error messages

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Do not set the identifier as an annotation

This is actually incorrect based on the OCI specification, the
annotation should be used only for program versions, i.e. application
version, but this is not avaiable to us at this point.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(oci): Be more verbose when setting KConfig (OSFeatures)

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(packmanager): Remove unused query option attributes

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(packmanager): Show only set (and previously unset) field attributes

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Speed up package manager initialization

Use internal KraftKit knowledge of the fact that the config often lists
the well-known path of the Manifest package manager's remote index.
This is obviously not an OCI image registry so we can safely skip it.
Doing this speeds up the kraft CLI and the instantiation of the OCI
Package Manager in general by a noticable amount, especially with
limited internet connectivity (as none is subsequently required).

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(push): Rework push to better handle package selection

This commit reworks the `kraft pkg push` subcommand to a). better handle
selection via the `--format` flag by iterating over package managers and
b). outputs relevant information package `Columns` to give a better
picture of what is being pushed (especially artifacts that have
identical names, i.e. OCI images).

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pull): Fix missing plat/arch/kconfig options to pull request

This commit fixes an issue where `kraft pkg pull` was a). not passing
the architecture and platform when positional arguments were set, such
that form:
  ```console
  kraft pkg pull --arch x86_64 unikraft.org/nginx:latest
  ```
Would pull all artifacts instead of by architecture; and b).
additionally setting KConfig options via the new `--kconfig|-k`
option(s) such that packages can be queried by individual configuration
parameters.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(pull): Add short-hand syntax for `--plat` flag

The user can use the shorthand `--plat qemu/x86_64` instead of setting
`--plat qemu --arch x86_64`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
docs(oci): Add initial `README.md`

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on cmd/kraft/build/build.go:
```diff
@@ -152,7 +152,7 @@ func (opts *Build) pull(ctx context.Context, project app.Application, workdir st
 					packmanager.WithName(opts.project.Template().Name()),
 					packmanager.WithTypes(unikraft.ComponentTypeApp),
 					packmanager.WithVersion(opts.project.Template().Version()),
-					packmanager.WithCache(!opts.NoCache),
```
shouldn't all of these also have an alias to `update` now? And then slowly deprecate `--no-cache`

nderjung replied:
Yes I agree. For now i didn't want to change any flags as this is more user facing. We can do this in a later step i think

craciunoiuc commented on oci/manifest.go:
```diff
@@ -28,7 +28,7 @@ import (
 )
 
-type Image struct {
+type Manifest struct {
```
I am not 100% sure I agree with this change because now `manifest.manifest` exists and can be confusing, but I'll not oppose it

nderjung replied:
I guess it could be called `ImageManifest` -- the OCI spec calls it this, or just "manifest". I agree it is a bit confusing but they are in different packages.

See https://github.com/opencontainers/image-spec/blob/main/manifest.md

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -65,6 +65,40 @@ func (handle *DirectoryHandler) DigestExists(ctx context.Context, dgst digest.Di
+	}
+
+	// Read the manifest
+	reader, err := os.Open(manifestPath)
```
`defer reader.Close()`

craciunoiuc commented on manifest/manager.go:
```diff
@@ -68,7 +72,22 @@ func NewManifestManager(ctx context.Context, opts ...any) (packmanager.PackageMa
 	if len(m.manifests) == 0 {
-		return nil, fmt.Errorf("no manifests specified in config")
+		// In this scenario, re-attempt all manifests provided within the config
+		// space which were note remotely probed during initialization.
```
`not`

craciunoiuc commented on oci/handler/containerd.go:
```diff
@@ -139,6 +143,135 @@ func (handle *ContainerdHandler) PullDigest(ctx context.Context, mediaType, full
+		err = combineErrors(err, done(ctx))
+	}()
+
+	writer, err := content.OpenWriter(
```
needs closing also `defer writer.close()`

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -306,6 +306,63 @@ func (handle *DirectoryHandler) PullDigest(ctx context.Context, mediaType, fullr
+		return fmt.Errorf("could not make parent directory: %w", err)
+	}
+
+	blob, err := os.OpenFile(blobPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o664)
```
needs closing  also `defer blob.Close()`

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -504,310 +704,278 @@ func (handle *DirectoryHandler) DeleteManifest(ctx context.Context, fullref stri
-	if err != nil {
-		return nil, err
-	}
+		indexFile, err := os.OpenFile(indexPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o664)
```
`defer indexFile.Close()`

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -504,310 +704,278 @@ func (handle *DirectoryHandler) DeleteManifest(ctx context.Context, fullref stri
 
-	// Read the config
-	reader, err = os.Open(configDir)
+	// Read the index
+	reader, err := os.Open(indexPath)
```
`defer reader.Close()`

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -504,310 +704,278 @@ func (handle *DirectoryHandler) DeleteManifest(ctx context.Context, fullref stri
-	}
-
-	writer, err = os.Create(configPath)
+	// Read the index
+	reader, err := os.Open(indexPath)
```
`defer reader.Close()`

craciunoiuc commented on oci/handler/directory.go:
```diff
@@ -504,310 +704,278 @@ func (handle *DirectoryHandler) DeleteManifest(ctx context.Context, fullref stri
 	}
 
-	image, err := handle.ResolveImage(ctx, fullref)
+	// Read the config
+	reader, err := os.Open(configDir)
```
`defer reader.Close()`

craciunoiuc commented on oci/pack.go:
```diff
@@ -248,144 +249,147 @@ func NewPackageFromTarget(ctx context.Context, targ target.Target, opts ...packm
-		return nil, fmt.Errorf("OCI image does not contain a Unikraft unikernel")
-	}
-
+// NewPackagrFromOCIManifestDigest is a constructor method which instantiates an
```
`a`

craciunoiuc commented on cmd/kraft/pkg/pkg.go:
```diff
@@ -164,13 +182,46 @@ func (opts *Pkg) Run(cmd *cobra.Command, args []string) error {
 	var tree []*processtree.ProcessTreeItem
 
-	parallel := !config.G[config.KraftKit](ctx).NoParallel
+	// parallel := !config.G[config.KraftKit](ctx).NoParallel
```
Should this be commented or just removed?

nderjung replied:
Left over artefact from debugging. I'll remove this commenting out.

craciunoiuc commented on cmd/kraft/pkg/pkg.go:
```diff
@@ -164,13 +182,46 @@ func (opts *Pkg) Run(cmd *cobra.Command, args []string) error {
+
+		// Set the merge strategy as an option that is then passed to the
+		// package manager.
+		case packmanager.StrategyMerge, packmanager.StrategyOverwrite:
```
Should this have a default case? Otherwise, should this be the default case?

nderjung replied:
Will set to `default`.

craciunoiuc commented on packmanager/strategy.go:
```diff
@@ -0,0 +1,67 @@
+	}
+}
+
+// MergeStrategyNames returns the string represention of all possible package
```
`representation`

craciunoiuc commented on cmd/kraft/pkg/pull/pull.go:
```diff
@@ -82,6 +83,12 @@ func (opts *Pull) Pre(cmd *cobra.Command, _ []string) error {
 	cmd.SetContext(ctx)
 
+	if strings.ContainsRune(opts.Platform, '/') && opts.Architecture == "" {
+		split := strings.SplitN(opts.Platform, "/", 2)
```
Also check if the split resulted into 2 strings, otherwise `qemu/` will result in a crash

craciunoiuc commented on oci/README.md:
```diff
@@ -0,0 +1,71 @@
+# Unikraft OCI Images 
+
+This package contains the implementation for handling both the packaging and
+distribution of Unikraft unikernels via the [Open Container Image (OCI)'s Image
```
Start a new line only on sentence end, one sentence per line.

nderjung replied:
)':

craciunoiuc reviewed (changes requested):
I left all comments inline. Also check the results from gofumpt.

nderjung committed:
fix(source): Request an update when checking source compatibility

Without requesting an update from the package manager, the compatibility
check can fail for sources which are external to the host.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung commented:
Git format check is failing because there are commit messages with URL citations that are longer than 74 characters.

craciunoiuc commented:
yup, noticed

I will now do a pass and try most of the build/run/pkg/pull/push arguments

nderjung committed:
fix(pkg): Handle packaging synchronously

To prevent any out-of-order package writes, resort to non-parallel
builds.  It is also not necessarily more efficient handle packaging in
parallel since files are always locked.  In turn, this also results in
the output of the packaging to read better.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pull): Use host plat/arch and rename flag `--all`

This commit first names the flag `--all-versions` to `--all`.  This
doesn't affect existing functionality since this flag was unused.
Instead, this flag is now used to determine whether to pull "everything"
from the selected catalog based on the provided name.  When unset, and
additionally when `--arch` and `--plat` are unset, the host is queried
for its architecture and platform, as a usability improvement.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Set the list of OS features (KConfig options)

Previously, the `NewManifestFromDigest` did not set the list of OS
features from the Image config which contains the list of KConfig
options for the image.  As a result, information was lost if the image
was re-serialized.  This commit fixes this by replacing the incorrectly
set `OSVariant` with `OSFeatures`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg)!: Switch `--with-kconfig` to `--no-kconfig`

Default to setting the KConfig for the image in the packaging process.
Ultimately, this information enriches the resulting artifact and helps
in determing the composition of the embedded unikernel, this should be
default functionality.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(multiselect): Introduce `multiselect` package

The `multiselect` package enables quick TUI-based prompts to narrow a
larger selection into a narrower selection.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(pkg): Add interactive prompts

This commit reworks `kraft pkg push` to use an interactive prompt to
select what to package before packing.  In addition, it uses the native
`target.Filter` method determine packages based on the provided CLI
flags for `--plat`, `--arch` and `--target`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(pkg): Make `--name` a mandatory flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat: Add `fmt.Stringer` to `unikraft.Nameable` interface

This method abstract the `unikraft.Nameable` interface to an even
simpler method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): Request an update before checking package compatibility

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Handle case where remote and local digests exist

In the scenario where a user has already pushed the same manifest to a
remote registry and still has the manifest locally, the `Catalog` method
would return both entries.  Instead, use the platform checksum as a
unique index in a map to prevent duplicates.  The local will be
preferred since this is queried last.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Only set authentication when registry is found

Previously, the `WithAuth` would have always been called with a nil
`authn.AuthConfig`.  This commit ensures that option method is only used
when authentication details are found.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(oci): Prefer KraftKit authentication details over docker

This commit incorporates two changes to the `defaultAuths` method.  The
first is to skip unset entries from `$HOME/.docker/config.json` and to
not populate those in the return set of auths.  The second is to prefer
KraftKit's authentication details over Dockers.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
What can go ~~wrong~~ right?

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
